File: clients/kotlin/synchro/src/test/kotlin/com/trainstar/synchro/SynchroConfigTests.kt
package com.trainstar.synchro

import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

class SynchroConfigTests {
    private val config = SynchroConfig(
        dbPath = "unused",
        serverURL = "http://localhost:8080",
        authProvider = { error("configuration must not request authentication") },
        clientID = "config-test",
        appVersion = "1.0.0",
    )

    @Test
    fun invalidSettingsFailDuringConstruction() {
        for (value in listOf(Double.NaN, Double.POSITIVE_INFINITY, Double.NEGATIVE_INFINITY, -1.0)) {
            assertThrows(IllegalArgumentException::class.java) { config.copy(syncInterval = value) }
            assertThrows(IllegalArgumentException::class.java) { config.copy(pushDebounce = value) }
        }
        for (value in listOf(-1, Int.MIN_VALUE)) {
            assertThrows(IllegalArgumentException::class.java) { config.copy(maxRetryAttempts = value) }
        }
        for (value in listOf(Int.MIN_VALUE, -1, 0, 1001, Int.MAX_VALUE)) {
            assertThrows(IllegalArgumentException::class.java) { config.copy(pullPageSize = value) }
            assertThrows(IllegalArgumentException::class.java) { config.copy(pushBatchSize = value) }
        }
    }

    @Test
    fun validLimitsPreserveManualSchedulingAndZeroRetries() {
        val manual = config.copy(syncInterval = 0.0, pushDebounce = 0.0, maxRetryAttempts = 0, pullPageSize = 1, pushBatchSize = 1)
        assertEquals(0.0, manual.syncInterval, 0.0)
        assertEquals(0.0, manual.pushDebounce, 0.0)
        assertEquals(0, manual.maxRetryAttempts)
        assertEquals(1, manual.effectivePullPageSize)
        val maximum = config.copy(pullPageSize = 1000, pushBatchSize = 1000, maxRetryAttempts = Int.MAX_VALUE)
        assertEquals(1000, maximum.effectivePullPageSize)
        assertEquals(1000, maximum.pushBatchSize)
        assertEquals(Int.MAX_VALUE, maximum.maxRetryAttempts)
    }
}

File: clients/swift/Tests/SynchroTests/SynchroConfigTests.swift
import XCTest
@testable import Synchro

final class SynchroConfigTests: XCTestCase {
    private func config(
        syncInterval: TimeInterval = 30,
        pushDebounce: TimeInterval = 0.5,
        maxRetryAttempts: Int = 5,
        pullPageSize: Int = 100,
        pushBatchSize: Int = 100
    ) -> SynchroConfig {
        SynchroConfig(
            dbPath: NSTemporaryDirectory() + "synchro_config_\(UUID().uuidString).sqlite",
            serverURL: URL(string: "http://config.invalid")!,
            authProvider: { XCTFail("Configuration checks must not request authentication"); return "" },
            clientID: "config-client",
            appVersion: "1.0.0",
            syncInterval: syncInterval,
            pushDebounce: pushDebounce,
            maxRetryAttempts: maxRetryAttempts,
            pullPageSize: pullPageSize,
            pushBatchSize: pushBatchSize
        )
    }

    func testInvalidLimitsAndTimersFailBeforeDatabaseCreation() async throws {
        var invalid = [
            config(maxRetryAttempts: -1),
            config(pullPageSize: 0), config(pullPageSize: 1001),
            config(pushBatchSize: 0), config(pushBatchSize: 1001),
        ]
        for timer in [-1, 0.0000000001, .nan, .infinity, -.infinity, Double.greatestFiniteMagnitude, Double(UInt64.max)] {
            invalid.append(config(syncInterval: timer))
            invalid.append(config(pushDebounce: timer))
        }
        for config in invalid {
            do {
                let client = try SynchroClient(config: config)
                try await client.close()
                XCTFail("Invalid configuration was accepted")
            } catch SynchroError.invalidResponse {
                XCTAssertFalse(FileManager.default.fileExists(atPath: config.dbPath))
            }
        }
    }

    func testZeroRetriesAndManualSchedulingRemainValid() async throws {
        for pageSize in [1, 1000] {
            let config = config(
                syncInterval: 0, pushDebounce: 0, maxRetryAttempts: 0,
                pullPageSize: pageSize, pushBatchSize: pageSize
            )
            let client = try SynchroClient(config: config)
            XCTAssertEqual(config.pullPageSize, pageSize)
            XCTAssertEqual(client.getSyncStatus(), .localReady)
            try await client.close()
        }
    }
}
